File: core/src/net/brenig/pixelescape/render/ui/ingame/AbilityWidget.java
package net.brenig.pixelescape.render.ui.ingame;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.ui.Button;
import com.badlogic.gdx.scenes.scene2d.ui.Skin;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.badlogic.gdx.scenes.scene2d.utils.Drawable;
import net.brenig.pixelescape.game.entity.impl.EntityPlayer;
import net.brenig.pixelescape.lib.Utils;
import net.brenig.pixelescape.screen.GameScreen;

/**
 * Gui Button that triggers use of the current player Ability
 */
public class AbilityWidget extends Button {


	private final GameScreen gameScreen;
	private EntityPlayer player;

	private float animCounter = 0;
	private final static float ANIM_DURATION = 0.5F;

	private final static float item_frame_border = 0.21875F;

	public AbilityWidget(Skin skin, String style, EntityPlayer player, GameScreen gameScreen) {
		this(skin.get(style, AbilityButtonStyle.class), player, gameScreen);
	}

	public AbilityWidget(Skin skin, EntityPlayer player, GameScreen gameScreen) {
		this(skin.get(AbilityButtonStyle.class), player, gameScreen);
	}

	public AbilityWidget(AbilityButtonStyle style, EntityPlayer player, GameScreen gameScreen) {
		super(style);
		this.gameScreen = gameScreen;
		setPlayer(player);
		initialize();
	}

	public AbilityWidget(EntityPlayer player, GameScreen gameScreen) {
		super();
		this.gameScreen = gameScreen;
		setPlayer(player);
		initialize();
	}

	/**
	 * initialized click listener etc.
	 */
	private void initialize() {
		addListener(new ClickListener() {
			@Override
			public void clicked(InputEvent event, float x, float y) {
				if (!gameScreen.isGamePaused() && player.hasAbility() && player.getCooldownRemaining() == 0) {
					player.useAbility();
				}
			}
		});
	}

	/**
	 * sets the player that has the ability
	 */
	public void setPlayer(EntityPlayer player) {
		this.player = player;
	}

	@Override
	public void draw(Batch batch, float parentAlpha) {
		super.draw(batch, parentAlpha);
		if (player.hasAbility()) {
			final float itemFrame = getWidth() * item_frame_border;
			final Drawable abilityIcon = player.getCurrentAbility().getDrawable(gameScreen.getGame().getGameAssets());
			abilityIcon.draw(batch, getX() + itemFrame, getY() + itemFrame, getWidth() - itemFrame * 2, getHeight() - itemFrame * 2);
			if (player.getCooldownRemaining() != 0) {
				animCounter = ANIM_DURATION;
				gameScreen.getGame().getRenderManager().setColor(0.7F, 0.7F, 1, 0.4F);
				gameScreen.getGame().getRenderManager().rect(batch, getX() + itemFrame, getY() + itemFrame, getWidth() - itemFrame * 2, (getHeight() - itemFrame * 2) * player.getCooldownRemainingScaled());
			} else if (animCounter > 0) {
				animCounter -= Gdx.graphics.getDeltaTime();
				final float alpha = Utils.easeInAndOut(animCounter, ANIM_DURATION) * 0.7F;
				gameScreen.getGame().getRenderManager().setColor(1, 1, 1, alpha);
				gameScreen.getGame().getRenderManager().rect(batch, getX() + itemFrame, getY() + itemFrame, getWidth() - itemFrame * 2, getHeight() - itemFrame * 2);
			}
		}
	}

	public static class AbilityButtonStyle extends Button.ButtonStyle {

		public AbilityButtonStyle() {

		}

		public AbilityButtonStyle(Drawable up, Drawable down, Drawable checked) {
			super(up, down, checked);
		}

		public AbilityButtonStyle(AbilityButtonStyle style) {
			super(style);
		}
	}
}

File: core/src/net/brenig/pixelescape/render/ui/ingame/AbilityWidget.kt
package net.brenig.pixelescape.render.ui.ingame

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.ui.Button
import com.badlogic.gdx.scenes.scene2d.ui.Skin
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import com.badlogic.gdx.scenes.scene2d.utils.Drawable
import net.brenig.pixelescape.game.entity.impl.EntityPlayer
import net.brenig.pixelescape.lib.utils.AnimationUtils
import net.brenig.pixelescape.render.PixelTextureRegion
import net.brenig.pixelescape.screen.GameScreen
import kotlin.math.min

/**
 * Gui Button that triggers use of the current player Ability
 */
class AbilityWidget : Button {


    private val gameScreen: GameScreen
    private val player: EntityPlayer


    private var animCounter = 0f

    constructor(skin: Skin, style: String, player: EntityPlayer, gameScreen: GameScreen) : this(skin.get<AbilityButtonStyle>(style, AbilityButtonStyle::class.java), player, gameScreen)

    constructor(skin: Skin, player: EntityPlayer, gameScreen: GameScreen) : this(skin.get<AbilityButtonStyle>(AbilityButtonStyle::class.java), player, gameScreen)

    constructor(style: AbilityButtonStyle, player: EntityPlayer, gameScreen: GameScreen) : super(style) {
        this.gameScreen = gameScreen
        this.player = player
        initialize()
    }

    /**
     * initialized click listener etc.
     */
    private fun initialize() {
        addListener(object : ClickListener() {
            override fun touchDown(event: InputEvent?, x: Float, y: Float, pointer: Int, button: Int): Boolean {
                if (!gameScreen.isGamePaused && player.hasAbility() && player.cooldownRemaining == 0f) {
                    player.useAbility()
                }
                return super.touchDown(event, x, y, pointer, button)
            }
        })
    }

    override fun draw(batch: Batch, parentAlpha: Float) {
        super.draw(batch, parentAlpha)
        if (player.hasAbility()) {
            val itemFrame = width * item_frame_border
            val abilityIcon = player.currentAbility!!.getDrawable(gameScreen.game.gameAssets)
            abilityIcon.draw(batch, x + itemFrame, y + itemFrame, width - itemFrame * 2, height - itemFrame * 2)
            if (player.cooldownRemaining != 0f) {
                animCounter = ANIM_DURATION
                gameScreen.game.renderManager.setColor(0.7f, 0.7f, 1f, 0.4f)
                gameScreen.game.renderManager.rect(batch, x + itemFrame, y + itemFrame, width - itemFrame * 2, (height - itemFrame * 2) * player.cooldownRemainingScaled)
            } else if (animCounter > 0) {
                animCounter -= Gdx.graphics.deltaTime
                val alpha = AnimationUtils.easeInAndOut(animCounter, ANIM_DURATION) * 0.7f
                gameScreen.game.renderManager.setColor(1f, 1f, 1f, alpha)
                gameScreen.game.renderManager.rect(batch, x + itemFrame, y + itemFrame, width - itemFrame * 2, height - itemFrame * 2)
            }

            if (player.remainingAbilityUses > 0) {
                val scale = min(player.remainingAbilityUses, 5).toFloat() / 5F
                style.usesBar?.draw(batch, x + charge_bar_x_pos * width, y + charge_bar_y_pos * height,
                        width * charge_bar_width * scale, height * charge_bar_height,
                        0, 0, (scale * charge_bar_texture_width.toFloat()).toInt(), charge_bar_texture_height)
            }
        }
    }
    override fun getStyle(): AbilityButtonStyle {
        return super.getStyle() as AbilityButtonStyle
    }

    class AbilityButtonStyle : Button.ButtonStyle {

        var usesBar: PixelTextureRegion? = null

        constructor() : super()

        constructor(up: Drawable, down: Drawable, checked: Drawable, usesBar: PixelTextureRegion?) : super(up, down, checked) {
            this.usesBar = usesBar
        }

        constructor(style: AbilityButtonStyle) : super(style)
    }

    companion object {
        private const val ANIM_DURATION = 0.5f

        private const val item_frame_border = 0.21875f
        private const val charge_bar_x_pos = 0.15625f
        private const val charge_bar_y_pos = 0.09375f
        private const val charge_bar_width = 0.625f
        private const val charge_bar_height = 0.0625f
        private const val charge_bar_texture_width = 20
        private const val charge_bar_texture_height = 2
    }
}
